packages/cdk8s/lib/app.ts

```diff
@@ -32,12 +37,118 @@ export class App extends Construct {
    * Synthesizes all manifests to the output directory
    */
   public synth(): void {
-    if (!fs.existsSync(this.outdir)) {
-      fs.mkdirSync(this.outdir);
+
+    fs.mkdirSync(this.outdir, { recursive: true });
+
+    // this is kind of sucky, eventually I would like the DependencyGraph
+    // to be able to answer this question.
+    const hasDependantCharts = resolveDependencies(this);
+
+    // Since we plan on removing the distributed synth mechanism, we no longer call `Node.synthesize`, but rather simply implement
+    // the necessary operations. We do however want to preserve the distributed validation.
+    validate(this);
+
+    const simpleManifestNamer = (chart: Chart) => `${Node.of(chart).uniqueId}.k8s.yaml`;
+    const manifestNamer = hasDependantCharts ? (chart: Chart) => `${index.toString().padStart(4, '0')}-${simpleManifestNamer(chart)}` : simpleManifestNamer;
+
+    const charts: IConstruct[] = new DependencyGraph(Node.of(this)).topology().filter(x => x instanceof Chart);
+
+    let index = 0;
+    for (const node of charts) {
+      const chart: Chart = Chart.of(node);
+      Yaml.save(path.join(this.outdir, manifestNamer(chart)), chartToKube(chart));
+      index++;
     }
 
-    Node.of(this).synthesize({
-      outdir: this.outdir
-    });
   }
+
+  /**
+   * Synthesize a single chart.
+   *
+   * Each element returned in the resulting array represents a different ApiObject
+   * in the scope of the chart.
+   *
+   * Note that the returned array order is important. It is determined by the various dependencies between
+   * the constructs in the chart, where the first element is the one without dependencies, and so on...
+   *
+   * @returns An array of JSON objects.
+   * @param chart the chart to synthesize.
+   * @internal
+   */
+  public static _synthChart(chart: Chart): any[] {
+
+    const app: App = App.of(chart);
+
+    // we must prepare the entire app before synthesizing the chart
+    // because the dependency inference happens on the app level.
+    resolveDependencies(app);
+
+    // validate the app since we want to call onValidate of the relevant constructs.
+    // note this will also call onValidate on constructs from possibly different charts,
+    // but thats ok too since we no longer treat constructs as a self-contained synthesis unit.
+    validate(app);
+
+    return chartToKube(chart);
+  }
+
+  private static of(c: IConstruct): App {
+
```

Comment:
Left this here since I don't think it clutters to much, and we might actually want to switch it to public at some point?